src/filters/WorldFilter.cpp

```diff
@@ -0,0 +1,129 @@
+        }
+    }
+
+//Creates a world message with the currently observed objects in it
+    proto::World WorldFilter::getWorld(double time) {
+        //First we update to the time we want packets at. Very important!
+        update(time, true);
+        proto::World world;
+        world.set_time(time);
+        for (const auto &kalmanYellowBotsOneId : yellowBots) {
+            if (!kalmanYellowBotsOneId.second.empty()) {
+                world.mutable_yellow()->Add(bestFilter(kalmanYellowBotsOneId.second)->asWorldRobot());
+            }
+        }
+        for (const auto &kalmanBlueBotsOneId : blueBots) {
+            if (!kalmanBlueBotsOneId.second.empty()) {
+                world.mutable_blue()->Add(bestFilter(kalmanBlueBotsOneId.second)->asWorldRobot());
+            }
+        }
+        if (!balls.empty()) {
+            proto::WorldBall worldBall = bestFilter(balls)->asWorldBall();
+            world.mutable_ball()->CopyFrom(worldBall);
+        }
+        return world;
+    }
+    void WorldFilter::update(double time, bool extrapolateLastStep) {
+        //TODO: Find a more pretty way to write a loop like this
+        const double removeFilterTime = 0.4; //Remove filters if no new observations have been added to it for this amount of time
+        updateRobots(yellowBots, time, extrapolateLastStep, removeFilterTime);
+        updateRobots(blueBots, time, extrapolateLastStep, removeFilterTime);
+        auto ball = balls.begin();
+        while (ball != balls.end()) {
+            ball->get()->update(time, extrapolateLastStep);
+            if (time - ball->get()->getLastUpdateTime() > removeFilterTime) {
+                balls.erase(ball);
+            } else {
+                ++ball;
+            }
+        }
+    }
+    void WorldFilter::updateRobots(robotMap &robots, double time, bool extrapolateLastStep, double removeFilterTime) {
+        for (auto &filtersAndId : robots) {
+            auto filter = filtersAndId.second.begin();
+            while (filter != filtersAndId.second.end()) {
+                filter->get()->update(time, extrapolateLastStep);
+                if (time - filter->get()->getLastUpdateTime() > removeFilterTime) {
+                    filtersAndId.second.erase(filter);
+                } else {
+                    ++filter;
+                }
+            }
+        }
+    }
+    const std::unique_ptr<RobotFilter> &
+    WorldFilter::bestFilter(const std::vector<std::unique_ptr<RobotFilter>> &filters) {
+        int bestIndex = 0;
+        int bestFrames = 0;
+        for (int i = 0; i < filters.size(); ++i) {
+            if (filters[i]->frames() > bestFrames) {
+                bestFrames = filters[i]->frames();
+                bestIndex = i;
+            }
+        }
+        return filters[bestIndex];
+    }
+    const std::unique_ptr<BallFilter> &
+    WorldFilter::bestFilter(const std::vector<std::unique_ptr<BallFilter>> &filters) {
+        int bestIndex = -1;
+        int bestFrames = -1;
```

Comment:
Could also be 0

Reply:
I would argue that's a bad idea. Filters are in the new system always constructed with a frame, so anything smaller than 1 is bad. The -1 is much more clearly problematic, whereas with 0 people might not realise that this is actually a problem.

Reply:
Then, in const std::unique_ptr<RobotFilter> &WorldFilter::bestFilter, 0 could be replaced with -1

Reply:
Oh you are talking about bestIndex not bestFrames! There's UB now, I see.